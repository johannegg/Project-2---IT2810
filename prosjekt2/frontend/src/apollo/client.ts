import { ApolloClient } from "@apollo/client";
import cache from "./cache";

const client = new ApolloClient({
	uri: "http://localhost:3001/", // Use this to run locally, for tests
	//uri: "http://it2810-12.idi.ntnu.no:3001", // Use this to run on virtual machine
	cache: cache,
});

export default client;
