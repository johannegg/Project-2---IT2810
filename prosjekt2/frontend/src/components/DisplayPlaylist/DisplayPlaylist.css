.playlist-details{
	display: flex;
	flex-direction: column;
	margin-top: 0;
	padding-top: 20px;
	position: absolute;
	top: 120px;
	left: 0;
	right: 0;
}

.playlist-details-container {
	display: flex;
	flex-direction: column;
	align-items: center;
	margin: 0 auto;
	min-width: 700px;
}

.playlist-details h1 {
	font-size: 2rem;
	margin-bottom: 25px;
	position: relative;
	align-self: center;
}

.playlist-details .songs-container {
	text-align: center;
}

.delete-button {
	background-color: #ea9ab2;
	color: white;
	border: none;
	padding: 10px;
	border-radius: 15px;
	cursor: pointer;
	width: 130px;
	text-align: center;
	align-self: flex-end;
}

.modal-overlay {
	position: fixed;
	top: 0;
	left: 0;
	width: 100%;
	height: 100%;
	background-color: rgba(0, 0, 0, 0.5);
	display: flex;
	justify-content: center;
	align-items: center;
	z-index: 1000;
}

.modal {
	background-color: white;
	padding: 20px;
	border-radius: 8px;
	box-shadow: 0 4px 8px rgba(0, 0, 0, 0.2);
	text-align: center;
	width: 300px;
}

.modal p {
	margin-bottom: 20px;
	font-size: 1.1rem;
}

.modal-buttons {
	display: flex;
	justify-content: center;
	gap: 20px;
}

.confirm-button {
	background-color: #ea9ab2;
	color: #000;
	border: none;
	padding: 10px 20px;
	border-radius: 5px;
	cursor: pointer;
}

.cancel-button {
	background-color: #ccc;
	color: #000;
	border: none;
	padding: 10px 20px;
	border-radius: 5px;
	cursor: pointer;
}

.confirm-button:hover {
	background-color: #d9829f;
}

.cancel-button:hover {
	background-color: #aaa;
}

@media (max-width: 850px) {
	.playlist-details h1 {
		font-size: 1.7rem;
	}
	.playlist-details-container {
		min-width: 330px;
	}

	.playlist-details{
		top: 90px;
	}
}

@media (max-width: 520px) {
	.playlist-details h1 {
		font-size: 1.5rem;
	}

	.delete-button {
		padding: 8px;
		width: 115px;
		font-size: 0.8rem;
	}
	.playlist-details{
		top: 80px;
	}
}


@media (prefers-color-scheme: dark) {
	.delete-button{
		background-color: var(--theme-favorite-pink);
	}
	.modal {
		background-color: rgb(54, 51, 51);
	}


}