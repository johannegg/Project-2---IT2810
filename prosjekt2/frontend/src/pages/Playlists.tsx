import React from "react";

const Playlists: React.FC = () => {
	return (
		<div>
			<h1>Your playlists</h1>
			<p>This is your playlists page.</p>
		</div>
	);
};

export default Playlists;
