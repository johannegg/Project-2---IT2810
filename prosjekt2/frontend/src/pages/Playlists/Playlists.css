.playlists-page {
    display: flex;
    flex-direction: column;
    margin-top: 0; 
    padding-top: 20px; 
    position: absolute; 
    top: 130px; 
    left: 0; 
    right: 0;
}

.new-playlist-button {
    background-color: #EA9AB2;
    color: white;
    border: none;
    padding: 10px;
    border-radius: 20px;
    cursor: pointer;
    width: 120px;
    text-align: center;
    margin-bottom: 20px; 
    font-size: 16px;
}


.playlists-container {
    grid-column: 1 / -1; 
    display: grid;
    grid-template-columns: repeat(4, 1fr);
    gap: 30px; 
    max-width: 1200px; 
    margin: 0 auto;
    margin-top: 20px; 
    margin-bottom: 20px;
}


@media (max-width: 1200px) {
    .playlists-container {
        grid-template-columns: repeat(3, 1fr); 
    }
}

@media (max-width: 1024px) {
    .playlists-container {
        grid-template-columns: repeat(2, 1fr); 
    }
}

@media (max-width: 600px) {
    .playlists-container {
        grid-template-columns: repeat(1, 1fr); 
    }
}


.playlist-card {
    border: 2px solid #000;
    border-radius: 12px;
    padding: 20px;
    width: 150px;
    height: 140px;
    text-align: center;
    cursor: pointer;
    transition: border-color 0.3s ease;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
}

.outer-playlist-container {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    max-width: 1200px;
    margin: 0 auto;
    margin-top: 20px;
}

