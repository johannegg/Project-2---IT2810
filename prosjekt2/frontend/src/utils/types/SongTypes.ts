export type Artist = {
    id: string;
    name: string;
};

export type Genre = {
    name: string;
};

export type SongData = {
    id: string;
    title: string;
    views: number;
    year: number;
    artist: Artist;
    genre: Genre;
    lyrics: string;
};