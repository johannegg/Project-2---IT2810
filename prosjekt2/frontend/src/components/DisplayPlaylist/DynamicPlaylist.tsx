import { useParams, useNavigate } from "react-router-dom";
import { useReactiveVar } from "@apollo/client";
import { playlistsVar } from "../../apollo/cache";
import DisplayPlaylist from "./DisplayPlaylist";
import { useEffect } from "react";

const DynamicPlaylist = () => {
	const { playlistId } = useParams<{ playlistId: string }>(); 
	const navigate = useNavigate();
	const playlists = useReactiveVar(playlistsVar);
	const playlistData = playlists.find((playlist) => playlist.id === playlistId);

	useEffect(() => {
		// Load playlists from localStorage into the Apollo cache when the component mounts
		const storedPlaylists = JSON.parse(localStorage.getItem("playlists") || "[]");
		playlistsVar(storedPlaylists);
	}, []);

	if (!playlistData) {
		// Redirect to a "not found" page if the playlist does not exist
		navigate("/not-found", { replace: true });
		return <div aria-label="Playlist not found">Playlist not found. Redirecting...</div>;
	}

	const handleDelete = () => {
		// Remove the playlist and update both the Apollo cache and localStorage
		const updatedPlaylists = playlists.filter((playlist) => playlist.id !== playlistId);
		playlistsVar(updatedPlaylists);
		localStorage.setItem("playlists", JSON.stringify(updatedPlaylists));
		// Navigate back to the playlists page after deletion
		if (playlistData) {
			navigate("/playlists", { state: { deletedPlaylistId: playlistData.id } });
		}
	};

	return (
		<section aria-label={`Dynamic view for playlist: ${playlistData.name}`}>
			{/* Render the DisplayPlaylist component for the current playlist */}
			<DisplayPlaylist playlistId={playlistData.id} onDelete={handleDelete} />
		</section>
	);
};

export default DynamicPlaylist;
