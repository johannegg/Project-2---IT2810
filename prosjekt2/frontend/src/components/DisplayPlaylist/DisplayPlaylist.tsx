import React from 'react';
import { PlaylistData } from '../../pages/Playlists/Playlists'; 
import './DisplayPlaylist.css';

interface DisplayPlaylistProps {
    playlist: PlaylistData;
}

const DisplayPlaylist: React.FC<DisplayPlaylistProps> = ({ playlist }) => {
    return (
        <div className="playlist-details">
            <h1>{playlist.name}</h1>
            <section className="songs-list">
                <h2>Songs in this playlist</h2>
                {playlist.songs.length > 0 ? (
                    <ul>
                        {playlist.songs.map((song) => (
                            <li key={song.id}>
                                {song.title} by {song.artist}
                            </li>
                        ))}
                    </ul>
                ) : (
                    <p>No songs available</p>
                )}
            </section>
        </div>
    );
};

export default DisplayPlaylist;
