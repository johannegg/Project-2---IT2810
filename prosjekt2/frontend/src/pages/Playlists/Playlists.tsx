import { useState, useEffect } from "react";
import "./Playlists.css";
import Playlist from "../../components/Playlist/Playlist";
import PlaylistForm from "../../components/PlaylistForm/PlaylistForm";
import { SongData } from "../../utils/types/SongTypes";
import { useNavigate } from "react-router-dom";
import { v4 as uuidv4 } from "uuid";

export interface PlaylistData {
	id: string;
	name: string;
	backgroundColor: string;
	icon: string;
	songs: SongData[];
}

const Playlists = () => {
	const navigate = useNavigate();

	const [playlists, setPlaylists] = useState<PlaylistData[]>(() => {
		const storedPlaylists = localStorage.getItem("playlists");
		return storedPlaylists ? JSON.parse(storedPlaylists) : [];
	});

	const [showForm, setShowForm] = useState(false);

	useEffect(() => {
		localStorage.setItem("playlists", JSON.stringify(playlists));
	}, [playlists]);

	const addNewPlaylist = (newPlaylistName: string, backgroundColor: string, icon: string) => {
		const newPlaylist = {
			id: uuidv4(),
			name: newPlaylistName,
			backgroundColor,
			icon,
			songs: [],
		};
		setPlaylists([...playlists, newPlaylist]);
	};

	const deletePlaylist = (playlistId: string) => {
		const updatedPlaylists = playlists.filter((playlist) => playlist.id !== playlistId);
		setPlaylists(updatedPlaylists);
	};

	const handlePlaylistClick = (playlist: PlaylistData) => {
		navigate(`/playlist/${playlist.id}`, { state: { playlist, deletePlaylist } });
	};

	return (
		<section className="playlists-page">
			<h1>Your Playlists</h1>

			<PlaylistForm show={showForm} onClose={() => setShowForm(false)} onSubmit={addNewPlaylist} />
			<div className="outer-playlist-container">
				<button onClick={() => setShowForm(true)} className="new-playlist-button">
					New Playlist
				</button>
				<div className="playlists-container">
					{playlists.length === 0 ? (
						<p>You have no playlists yet.</p>
					) : (
						playlists.map((playlist) => (
							<Playlist
								id={playlist.id}
								key={playlist.id}
								name={playlist.name}
								backgroundColor={playlist.backgroundColor}
								icon={playlist.icon}
								songs={playlist.songs}
								onClick={() => handlePlaylistClick(playlist)}
								tabIndex={0}
								onKeyDown={(e) => {
									if (e.key === "Enter" || e.key === " ") {
										handlePlaylistClick(playlist);
									}
								}}
							/>
						))
					)}
				</div>
			</div>
		</section>
	);
};

export default Playlists;
