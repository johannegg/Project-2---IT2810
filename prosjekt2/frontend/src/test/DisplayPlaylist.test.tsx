import { describe, it, expect, vi } from "vitest";
import { act, fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ApolloProvider, InMemoryCache, ApolloClient } from "@apollo/client";
import { PlaylistData } from "../pages/Playlists/Playlists";
import DisplayPlaylist from "../components/DisplayPlaylist/DisplayPlaylist";
import { makeVar } from "@apollo/client";
import { Artist, Genre } from "../utils/types/SongTypes";

const mockPlaylistsVar = makeVar<PlaylistData[]>([]);

vi.mock("@apollo/client", async (importOriginal) => {
	const actual = (await importOriginal()) as Record<string, unknown>;
	return {
		...actual,
		useReactiveVar: vi.fn(() => mockPlaylistsVar()),
		makeVar: actual.makeVar,
	};
});

// Create a mock Apollo Client
const mockApolloClient = new ApolloClient({
	uri: "http://localhost:4000", // Mock URL (not used in testing)
	cache: new InMemoryCache(),
});

const renderWithProviders = (component: React.ReactNode) => {
	return render(
		<ApolloProvider client={mockApolloClient}>
			<MemoryRouter>{component}</MemoryRouter>
		</ApolloProvider>,
	);
};

describe("DisplayPlaylist Component", () => {
	it("renders correctly and matches snapshot", () => {
		const mockArtist: Artist = { id: "a1", name: "Artist 1" };
		const mockGenre: Genre = { name: "pop" };

		const mockPlaylist: PlaylistData = {
			id: "1",
			name: "My Playlist",
			backgroundcolor: "#ffffff",
			icon: "ðŸŽµ",
			songs: [
				{
					id: "s1",
					title: "Song 1",
					views: 1000,
					year: 2020,
					artist: mockArtist,
					genre: mockGenre,
					lyrics: "Some lyrics for song 1",
				},
			],
		};

		act(() => {
			mockPlaylistsVar([mockPlaylist]);
		});

		const { asFragment } = renderWithProviders(
			<DisplayPlaylist playlistId="1" onDelete={vi.fn()} />,
		);

		expect(asFragment()).toMatchSnapshot();
	});

	it('displays "Playlist not found" when no playlist matches', () => {
		act(() => {
			mockPlaylistsVar([]); // Set an empty playlist array
		});

		renderWithProviders(<DisplayPlaylist playlistId="invalid" onDelete={vi.fn()} />);

		expect(screen.getByText("Playlist not found")).toBeInTheDocument();
	});

	it("calls onDelete and closes modal when confirmed", () => {
		const mockOnDelete = vi.fn();

		const mockPlaylist: PlaylistData = {
			id: "1",
			name: "My Playlist",
			backgroundcolor: "#ffffff",
			icon: "ðŸŽµ",
			songs: [
				{
					id: "s1",
					title: "Song 1",
					views: 1000,
					year: 2020,
					artist: { id: "a1", name: "Artist 1" },
					genre: { name: "pop" },
					lyrics: "Some lyrics for song 1",
				},
			],
		};

		act(() => {
			mockPlaylistsVar([mockPlaylist]);
		});

		renderWithProviders(<DisplayPlaylist playlistId="1" onDelete={mockOnDelete} />);

		const deleteButton = screen.getByRole("button", { name: /Delete playlist/i });
		fireEvent.click(deleteButton);

		expect(screen.getByText(/are you sure you want to delete this playlist/i)).toBeInTheDocument();

		const confirmButton = screen.getByText(/yes/i);
		fireEvent.click(confirmButton);

		expect(mockOnDelete).toHaveBeenCalled();

		expect(
			screen.queryByText(/are you sure you want to delete this playlist/i),
		).not.toBeInTheDocument();
	});
});
