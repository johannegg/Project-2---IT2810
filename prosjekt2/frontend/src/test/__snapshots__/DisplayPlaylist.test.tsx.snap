// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DisplayPlaylist Component > renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <section
    aria-label="Details for playlist My Playlist"
    class="playlist-details"
  >
    <div
      class="playlist-details-container"
    >
      <header
        class="playlist-header"
      >
        <button
          class="back-button"
        >
          ‚ùÆ Go back
        </button>
        <button
          aria-label="Delete playlist"
          class="delete-playlist-button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
            />
          </svg>
        </button>
      </header>
      <h1
        aria-label="Playlist name: My Playlist"
      >
        My Playlist üéµ
      </h1>
      <div
        aria-label="Songs in My Playlist"
        class="songs-container"
      >
        <section
          class="songContainer"
        >
          <table
            class="songTable"
          >
            <tbody>
              <tr
                class="tableRow"
                tabindex="0"
              >
                <td
                  class="title-artist-cell"
                >
                  <span
                    class="titleCell"
                  >
                    Song 1
                  </span>
                  <span
                    class="artistCell"
                  >
                    Artist 1
                  </span>
                </td>
                <td>
                  2020
                </td>
                <td
                  class="viewsCell"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    style="margin-right: 5px;"
                    viewBox="0 0 576 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
                    />
                  </svg>
                  1K
                </td>
                <td
                  class="plusMinusCell"
                >
                  <button
                    class="plusMinus-button"
                    data-label="Remove song"
                  >
                    <svg
                      color="#afc188"
                      fill="none"
                      height="1em"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      style="color: rgb(175, 193, 136);"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <circle
                        cx="12"
                        cy="12"
                        r="10"
                      />
                      <line
                        x1="8"
                        x2="16"
                        y1="12"
                        y2="12"
                      />
                    </svg>
                  </button>
                </td>
                <td>
                  <button
                    class="favoriteButton"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-heart "
                      data-icon="heart"
                      data-prefix="far"
                      focusable="false"
                      role="img"
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M225.8 468.2l-2.5-2.3L48.1 303.2C17.4 274.7 0 234.7 0 192.8l0-3.3c0-70.4 50-130.8 119.2-144C158.6 37.9 198.9 47 231 69.6c9 6.4 17.4 13.8 25 22.3c4.2-4.8 8.7-9.2 13.5-13.3c3.7-3.2 7.5-6.2 11.5-9c0 0 0 0 0 0C313.1 47 353.4 37.9 392.8 45.4C462 58.6 512 119.1 512 189.5l0 3.3c0 41.9-17.4 81.9-48.1 110.4L288.7 465.9l-2.5 2.3c-8.2 7.6-19 11.9-30.2 11.9s-22-4.2-30.2-11.9zM239.1 145c-.4-.3-.7-.7-1-1.1l-17.8-20-.1-.1s0 0 0 0c-23.1-25.9-58-37.7-92-31.2C81.6 101.5 48 142.1 48 189.5l0 3.3c0 28.5 11.9 55.8 32.8 75.2L256 430.7 431.2 268c20.9-19.4 32.8-46.7 32.8-75.2l0-3.3c0-47.3-33.6-88-80.1-96.9c-34-6.5-69 5.4-92 31.2c0 0 0 0-.1 .1s0 0-.1 .1l-17.8 20c-.3 .4-.7 .7-1 1.1c-4.5 4.5-10.6 7-16.9 7s-12.4-2.5-16.9-7z"
                        fill="currentColor"
                      />
                    </svg>
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </section>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
